### PR TITLE
ASYNC implementation for Presage predicitons

### LINKED ISSUE #9: Predictions using async API
I have started looking into presage, into it's internals and performance. There is a lot it has to do to get predictions and while I do have ideas on how to make it faster, I suggest to make predictions in separate thread. Namely, QML interface would add changes in context, keyboard processing as it has now, but would get prediction via signal (or just notified via signal that prediction changed). 

Such approach would allow us to later use more advanced language models, additional predictors and keep fast response of the keyboard .

### PR DESCRIPTION
**Pull request by rinigus**
This PR fixes https://github.com/martonmiklos/sailfishos-presage-predictor/issues/9 by pushing predictions into separate thread.

This PR makes a separation between linking with QML in GUI thread and Presage predictor. Presage object is created in the constructor of PresagePredictor and put by it into separate thread using Presage wrapper PresageWorker. Communication between  PresagePredictor and PresageWorker is implemented by queued signals allowing to perform predictions in asynchronous manner. 

Here,  PresagePredictor is also used as a stack which asks for new prediction and, when the worker thread is ready to do so, gives the latest request. Thus, if the worker will fall behind, it will be able to skip few predictions that are not needed anymore and perform the latest one. For this to work, language and "past buffers" are handled accordingly. 

As before, RPMs are available are at https://build.merproject.org/package/show/home:rinigus:keyboard/sailfishos-presage-predictor . For ARM devices, grab http://repo.merproject.org/obs/home:/rinigus:/keyboard/sailfish_latest_armv7hl/armv7hl/maliit-plugin-presage-1.0+async.20180227153247.ab6a779-10.22.1.jolla.armv7hl.rpm and install on the top of Marisa-based RPM to get the latest version.

### CODE REVIEW

File: src/presagepredictor.cpp [hunk 1]
```diff
@@ -99,28 +124,31 @@ void PresagePredictor::predict()
...
 
 void PresagePredictor::acceptPrediction(int index)
 {
-    if (m_presage == nullptr)
-        return;
     log(QString("PresagePredictor::acceptPrediction(%1)").arg(index));
-    if ((size_t)index < m_predictedWords.size()) {
-        m_wordBuffer.clear();
+    if (index < m_predictedWords.size()) {
+        {
```
Comment (martonmiklos) [inline src/presagepredictor.cpp:148]: What is the reason for these braces? I am just curious.

File: src/presagepredictor.cpp [hunk 2]
```diff
@@ -75,17 +84,33 @@ void PresagePredictor::predict()
...
-    for(std::vector<std::string>::const_iterator i = m_predictedWords.begin(); i != m_predictedWords.end(); ++i)
-        words.append(QString::fromStdString(*i));
-    m_engine->reload(words);
+}
+
+bool PresagePredictor::contextStream(size_t &id, QString &language, std::string &buffer)
+{
+    QMutexLocker _(&m_mutex);
+
+    if (id == m_prediction_id)
```
Comment (martonmiklos) [inline src/presagepredictor.cpp:104]: I would add a brace pair here just for mistake proofing.

File: src/presagepredictor.h [hunk 3]
```diff
@@ -78,45 +83,65 @@ class PresagePredictor : public QQuickItem
...
+
     PresagePredictorModel *engine() const;
 
 protected:
 
-    QStringList predictions() const;
-    void setPredictions(const QStringList &predictions);
-
-    void predict();
+    void predict(); ///< Request new prediction. NB! Call only with the locked mutex
```
Comment (martonmiklos) [inline src/presagepredictor.h:103]: Just as curiosity: I have seen this NB! in several comments what does it mean?

File: src/presagepredictor.cpp (hunk 1 above)
Comment (rinigus) [reply src/presagepredictor.cpp:148]: This was to protect the mutex. I wasn't sure whether `setEngineCapitalization` could in future lead to locking of the mutex. So, I kept its action local by the braces.

File: src/presagepredictor.cpp (hunk 2 above)
Comment (rinigus) [reply src/presagepredictor.cpp:104]: Here the mutex has to protect during full function execution. This function has to return consistent language and text and these cannot be changed meanwhile. This function is called from the worker and we are preventing changes in language or context while its executing. So, here we cannot separate the part to make braces around.

File: src/presagepredictor.h (hunk 3 above)
Comment (rinigus) [reply src/presagepredictor.h:103]: NB - https://en.wikipedia.org/wiki/Nota_bene :)

Comment (martonmiklos) [reply src/presagepredictor.h:103]: Ah thanks, I learned something new again.

**Comment by martonmiklos** [comment]
Nice job!

**Comment by rinigus** [comment]
Thank you very much! It was mainly moving your code around :)

**Pull request merged**